### PR TITLE
refactor: extract structs for downloading tarballs and npm registry packuments

### PR DESCRIPTION
**Pull request by dsherret**
This is part of the work for https://github.com/denoland/deno/issues/24063 where we need to ensure npm installation works properly on many workers.

This extracts out two structs that can be shared amongst the workers. One for downloading and extracting npm tarballs and another for downloading npm registry packuments.

### CODE REVIEW

File: cli/npm/managed/cache/registry_info.rs [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+      npmrc,
+      progress_bar,
+      memory_cache: Default::default(),
+    }
+  }
+
+  pub async fn load_package_info(
+    &self,
+    name: &str,
+    current_runtime_http_client: &Arc<HttpClient>,
```
Comment (dsherret) [inline cli/npm/managed/cache/registry_info.rs:81]: Part of this refactor is working towards ensuring we're not sharing HttpClients amongst different runtimes. I'll make this an `Rc<HttpClient>` in the future.

File: cli/npm/managed/mod.rs [hunk 2]
```diff
@@ -246,15 +254,17 @@ async fn snapshot_from_lockfile(
 /// An npm resolver where the resolution is managed by Deno rather than
 /// the user bringing their own node_modules (BYONM) on the file system.
 pub struct ManagedCliNpmResolver {
-  api: Arc<CliNpmRegistryApi>,
   fs: Arc<dyn FileSystem>,
   fs_resolver: Arc<dyn NpmPackageFsResolver>,
-  global_npm_cache: Arc<NpmCache>,
-  resolution: Arc<NpmResolution>,
+  http_client: Arc<HttpClient>,
```
Comment (dsherret) [inline cli/npm/managed/mod.rs:259]: I'll work on ensuring this is in an ideal place later on. Just didnt' want to land a massive refactor in one go.

File: tests/testdata/npm/deno_run_non_existent.out [hunk 3]
```diff
@@ -1,3 +1,2 @@
 Download http://localhost:4260/mkdirp
-Download http://localhost:4260/mkdirp
```
Comment (dsherret) [inline tests/testdata/npm/deno_run_non_existent.out:2]: It downloading this multiple times was a cosmetic bug previously from https://github.com/denoland/deno/pull/18622/. We already got the latest information from the npm registry, so we don't need to download it again (this is now enforced by the `RegistryInfoDownloader`)

File: cli/npm/managed/cache/registry_info.rs [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+use deno_npm::registry::NpmPackageInfo;
+
+use crate::args::CacheSetting;
+use crate::http_util::HttpClient;
+use crate::npm::common::maybe_auth_header_for_npm_registry;
+use crate::util::progress_bar::ProgressBar;
+
+use super::NpmCache;
+
+// todo(dsherret): create seams and unit test this
```
Comment (bartlomieju) [inline cli/npm/managed/cache/registry_info.rs:28]: What does "seams" mean in this context?

**Review by bartlomieju — APPROVED** [review]

File: cli/npm/managed/cache/registry_info.rs (hunk 4 above)
Comment (dsherret) [reply cli/npm/managed/cache/registry_info.rs:28]: A seam is a place in the code where you can modify the behaviour for testing without modifying the code. I believe it's a term coinded by Michael Feathers from the book "working effectively with legacy code" (probably one of the best books I've read). I kind of view it as a general term for a bunch of different techniques used for making code more testable.

**Pull request merged**